File: 0001-0500/0028/0028.find-the-index-of-the-first-occurrence-in-a-string.js
/*
 * @lc app=leetcode id=28 lang=javascript
 *
 * [28] Find the Index of the First Occurrence in a String
 */

// @lc code=start
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function (haystack, needle) {
    const l1 = haystack.length;
    const l2 = needle.length;

    if (l2 > l1) {
        return -1;
    }

    for (let i = 0; i <= l1 - l2; i++) {
        let j;
        for (j = 0; j < l2; j++) {
            if (haystack[i + j] !== needle[j]) {
                break;
            }
        }
        if (j === l2) {
            return i;
        }
    }

    return -1;
};
// @lc code=end

File: 0001-0500/0028/0028.find-the-index-of-the-first-occurrence-in-a-string.cpp
/*
 * @lc app=leetcode id=28 lang=cpp
 *
 * [28] Find the Index of the First Occurrence in a String
 */

// @lc code=start
class Solution
{
public:
    int strStr(string haystack, string needle)
    {
        if (needle.empty())
            return 0;
        int n = haystack.size(), m = needle.size();
        for (int i = 0; i < n - m + 1; ++i)
        {
            int j = 0;
            for (; j < m; ++j)
            {
                if (haystack[i + j] != needle[j])
                    break;
            }
            if (j == m)
                return i;
        }
        return -1;
    }
};
// @lc code=end
